### ISSUE TITLE
feature/question: cutsel agent with ability to remove separators

### ISSUE THREAD

**Issue by likebooks** [scipopt/PySCIPOpt#827]
Hi there! First of all thanks for supporting the python interface.

I am trying to implement a cutsel plugin with the following workflow:
1. Obtains `new_potential_cuts`: Gets generated cuts (list of SCIP Rows) _this is already implemented as an input of the cutselect method_
2. Obtains `previous_active_cuts`: Gets other cuts active in the LP that have been added previously (list of SCIP Rows): From looking at the SCIP docs I see [SCIPgetCuts](https://scipopt.org/doc/html/group__PublicCutMethods.php#ga957ce862ee379dcf71c6fbb647807c1f).
3. Applies a re-selection criteria to `new_potential_cuts` and `previous_active_cuts` so that:
3.1. A subset of `new_potential_cuts` is added _this is already implemented as output of the cutselect method_
3.2. A subset of `previous_active_cuts` is removed: from looking at the SCIP docs I see [sepastoreDelCut](https://scipopt.org/doc/html/sepastore_8c.php#a921cb191c7ecfc6bcc07c4b92f865733).

Some questions:
* In a C implementation, would `SCIPgetCuts` + `sepastoreDelCut` inside a `Cutsel` class be the proper way to implement this?
* From looking at the [PySCIPOpt docs](https://scipopt.github.io/PySCIPOpt/docs/html/classpyscipopt_1_1scip_1_1Model.html) I am only able to find methods to add/delete constraints [which I read are treated differently than separators](https://scipopt.org/doc/html/FAQ.php#conshdlrvsseparator) and for getting cuts I am only able to find method to retrieve the number of cuts. Any indications on how could I achieve the previous workflow with the current PySCIPOpt interface? It would be very convenient for me to be able to work in python as the rest of the environment I have is in python :)

Thanks!

(Edit i): Follow-up on this: I managed to extend the `SCIPgetCuts` method but I am having trouble with the `sepastoreDelCut`. 

After looking at the 'c' api it is seems that this method is only accessed through the [SCIPsepastoreRemoveInefficaciousCuts](https://www.scipopt.org/doc-6.0.1/html/sepastore_8c_source.php#l01009) method which is also only accessed through the [SCIPremoveInefficaciousCuts](https://www.scipopt.org/doc-6.0.1/html/group__PublicCutMethods.php#ga44b0ec6233e1c5d4edb3313c119ab0ee). 

Such method does not seem to have dynamic efficiency criterias:
![image](https://github.com/scipopt/PySCIPOpt/assets/164537094/b0b30333-c3c4-49be-97fe-2bacecad985c)

(Edit ii): [pyscipopt - SCIP: How to remove cuts from the sepastore under a user defined scoring criteria - Stack Overflow](https://stackoverflow.com/questions/78215851/scip-how-to-remove-cuts-from-the-sepastore-under-a-user-defined-scoring-criteri)

**Comment by likebooks** [comment]
Follow-up on this: I managed to extend the `SCIPgetCuts` method but I am having trouble with the `sepastoreDelCut`. 

After looking at the 'c' api it is seems that this method is only accessed through the [SCIPsepastoreRemoveInefficaciousCuts](https://www.scipopt.org/doc-6.0.1/html/sepastore_8c_source.php#l01009) method which is also only accessed through the [SCIPremoveInefficaciousCuts](https://www.scipopt.org/doc-6.0.1/html/group__PublicCutMethods.php#ga44b0ec6233e1c5d4edb3313c119ab0ee). 

Such method does not seem to have dynamic efficiency criterias:
![image](https://github.com/scipopt/PySCIPOpt/assets/164537094/b0b30333-c3c4-49be-97fe-2bacecad985c)

**Comment by likebooks** [comment]
[pyscipopt - SCIP: How to remove cuts from the sepastore under a user defined scoring criteria - Stack Overflow](https://stackoverflow.com/questions/78215851/scip-how-to-remove-cuts-from-the-sepastore-under-a-user-defined-scoring-criteri)

**Comment by Opt-Mucca** [comment]
Heyho! It's very cool to see people using the cut selection interface creatively! Nice question. 

In general, deleting applied cuts is non-trivial. I would need to ask some colleagues to confirm, but I don't think there's a simple function to remove them from the LP. SCIP essentially self-manages the cuts by removing those that are redundant after some threshold of LP solves. In case you haven't seen this information as well: https://www.scipopt.org/doc/html/FAQ.php#sepastorecutpool That link should describe the difference between sepastore and a cutpool. The actual answers:

- That implementation wouldn't be correct. `sepastoreDelCut` removes potential cuts from the separation storage of an upcoming separation round. It will not remove an already applied cut that does not exist in the storage currently. 
- One could always try hacking their way around this problem. You can relax the cuts that you want to delete, e.g. by increasing their RHS values or just making all their coefficients 0 etc. This is a bit wasteful, and I haven't tested if it works nicely, but it should ultimately remove the cut from the LP at some later LP solve for being redundant. As it's redundant it should also not affect the optimal solution value of the next LP solve (might affect the number of pivots). 

How did you extend the `SCIPgetCuts`? Was it iterating over the LP rows and checking if they are derived from a separator?

**Comment by likebooks** [comment]
Thanks for the detailed response!

i) I see, I didn't understand this properly :)
ii) This looks like a simple enough workaround. The original motivation was in reducing the number of pivots generated by low score cuts (based on the criteria) so maybe this is an ok performing option but not semantically strict to what I was aiming to test. I am trying to hack something using local node constraints which looks quite more wasteful but at the moment my main goal is to evaluate based on number of rounds (so I can take the time perf loss I guess).

a) My bad, extending was not the proper word I just added it to the python interface.

If you and your colleagues can point a direction for a proper way for cut removal with dynamic efficiency criterias in SCIP without having to hack around I would be very interested in contributing :):) (to either the python and C api's). For now, I think that one of the more hacky options will do the work.

**Comment by Opt-Mucca** [comment]
I just talked to a colleague: We concluded that SCIP is probably protecting the user by not providing this functionality. I"m not sure how warm-starting the simplex would be affected by relaxing a currently active cut for instance. I'm also not sure if an error would be raised when your new set of cuts somehow has a worse bound than the older LP (in the case that you removed some good cuts incorrectly).

I will wait for some else to respond if they think there's a a better workaround

**Comment by likebooks** [comment]
Hi @Opt-Mucca, I am trying to test the "new set of cuts somehow has a worse bound than the older LP".

In order to play around with this I would like to, inside the cutsel agent:
* get a fresh LP equivalent to the current LP in the given node where the cutsel agent was called.
* add/ remove cuts.
* solve the LP and get the bound with the cuts.

Is there a clean way to do this? I tried declarating a new LP() object, iterating over the linear constraints obtained with scip.getConss() and adding them as rows + adding the objective value with addCols() but it feels a bit chunky and I sometimes get a bound which is smaller than the previous I had which if done properly should not be possible.

**Comment by Opt-Mucca** [comment]
This is exactly what dive mode is made for. Just call `scip.startDive()` and then use all the diving functionality that you want e.g. `scip.addRowDive()` and `scip.solveDiveLP()`. Make sure to call `scip.endDive()` after trying all these changes. This saves you declaring a new LP object, and allows all the code to be nicely written inside of the cut selection routine.

**Comment by likebooks** [comment]
I have a follow-up question, for this mode is there a straightforward way to get the objective value (or the solution x) in the original variable space? After calling `scip.solveDiveLP()` with just added cuts the `scip.getLPObjVal()` gives me a very large value (from seeing the function docs I assume this is because its not in the original space), the alternative `scip.getObjVal()` gives me the previous solution value which I already had. I found `getLPSol` but it is defined variable-wise...

Thanks for the patience and supporting experimentation :):)

**Comment by Opt-Mucca** [comment]
I would assume that `scip.getLPObjVal()` works in dive mode. There shouldn't be any additional changes to the transformed space either, so there shouldn't be huge jumps in LP objective values between separation rounds (depends on the cuts of course). I'd encourage you to test that by explicitly calculating the objective and comparing (you can explicitly calculate the objective by getting the objective coefficients and LP solution value for each variable). You shouldn't use `scip.getObjVal()` because that uses the current primal solution, not the current LP solution.

**Comment by likebooks** [comment]
Thanks @Opt-Mucca this worked! After playing a bit with the basic methods, I am curious on why [addCons()](https://scipopt.github.io/PySCIPOpt/docs/html/classpyscipopt_1_1scip_1_1Model.html#aaeb2227123715120e5921b6633750644) asserts that the constraint parameter is an ExprCons. I would like to also pass "pyscipopt.scip.Constraints" to it. Is there a fast way to do this without manually iterating over the coefficients?

**Comment by Opt-Mucca** [comment]
The `ExprCons` is there as people traditionally build a `Expr` when they construct the constraint through Python. `Expr` is a class that ensures the appropriate expression is constructed. For example, when you square a variable, take `sin()` of term, or add a monomial. `ExprCons` is just an expression with a LHS and RHS. 

There is currently no fast way to pass a constraint to the `addCons` function. As you said, one would need to manually iterate over the coefficients. This shouldn't be difficult in the linear case though. 

I think cases like yours are quite fringe, as I assume you instantly want to then change the new created constraint. This would be trivial in the original model construction stage as you could just store the expression. Adding a `getExpr` function for a `scip.Constraint` would be a helpful addition to the repository, although likely is non-trivial.

**Comment by likebooks** [comment]
Thanks once and once again @Opt-Mucca!! Iterating over the coefficients worked as you mentioned! I made a utils function for the linear case, maybe once I am done with implementing this idea I can try and send a PR for the general getExpr case... might be a fun exercise :)

There is one more issue though, I changed my problems from minimization to a maximization objective but I am experiencing a strange behaviour in diving mode. I have a function that (0) takes the initial LP objective `LP_init`, then starts a dive where: 1) adds some cuts, 2) re-solves the problem 3) gets the latest LP objective value `LP_latest`. As we have only added constraints the region at (3) should be contained by the region at (0) -> `LP_init` >= `LP_latest` for a **maximization objective**. But I am experiencing the opposite (`LP_init` is somehow < `LP_latest`). Any ideas on why could this be happening? With minimization it worked as expected (opposite case so `LP_init` <= `LP_latest`). Is minimization assumed somewhere and should be now strictly specified?
